packages/flutter/lib/src/foundation/collections.dart

```diff
@@ -4,16 +4,16 @@
 
 // TODO(ianh): These should be on the Set and List classes themselves.
 
-/// Compares two sets for deep equality.
+/// Compares two sets for shallow equality.
```

Comment:
Reading this, I would assume "shallow" equality was just `list1 == list2`, and not comparing any of the elements.

Maybe use "element by element equality"?

Reply:
Hi @gspencergoog, Thank you for your review. I think the phrase "shallow equality" means exactly what is needed here since it refers to the equality of the objects instead of references (`list1 == list2`). But only in the first level in contrast to deep equality that checks recursively.

Element-by-element equality is not bad; however, it doesn't indicate that the operation is recursive or not (deep or shallow).

What's your opinion about "Compares two sets for shallow (element-by-element) equality."

Reply:
Or maybe "Compares two sets for shallow equality (element-by-element)."

Reply:
Whether it is deep or shallow isn't actually known here, though: if each element's operator== uses its own element-by-element comparison, then it actually is a deep comparison. If it just uses the identity comparison, then it's shallow. That's why I think you should probably avoid both deep and shallow and just describe what it uses (element by element).

Reply:
I understand your point. However, the user who reads this documentation should find out the output if they call these functions for a collection. No one is expected to know the implementation of operator== of the built-in collections. By the way, is it possible that the implementation of operator== will change without rewriting these functions? (There's a TODO comment that says these functions should be on collection classes themselves, so I think a bigger refactoring that could affect the implementation of operator== also would eliminate these functions.)

Reply:
If you use the term "shallow", then people will be surprised if they put composite objects into the list and get a deep comparison (because those composites might implement deep comparisons). Which means they actually will be required to know the implementation of `operator==` on the elements put into the list. If you use the term "element by element", then they will not be surprised, because you told them exactly what the behavior is. I think both "shallow" and "deep" here are misleading.

As for whether the implementations of `operator==` can change without `listEquals` changing: Yes they can change without rewriting these functions. The `listEquals` function takes a `List<T>`, and T can be anything, so if the type is a composite object (an object that contains other objects), and that type implements an `operator==` that compares all of its children with the "other" object's children, then you'll get deep equality. The `listEquals` function doesn't know this, and has no way to ask. It's even possible for some parts of the list to be compared deeply and others shallowly.